### PR TITLE
Allow building with clang-cl

### PR DESCRIPTION
**Pull request by MisterDA**
> [clang-cl](https://clang.llvm.org/docs/UsersManual.html#clang-cl) is an alternative command-line interface to Clang, designed for compatibility with the Visual C/C++ compiler, `cl.exe`.

clang-cl is ABI and (mostly) command-line compatible with cl, providing a drop-in replacement. Users may prefer it to MSVC as it's based on [LLVM/clang](https://llvm.org/), a free software compiler toolchain,

> which aims to deliver amazingly fast compiles, extremely useful error and warning messages and to provide a platform for building great source level tools.

I've added a job to the CI matrix to test OCaml with clang-cl on Windows x86_64. To test this PR locally, follow [Clang/LLVM support in Visual Studio projects](https://learn.microsoft.com/en-us/cpp/build/clang-support-msbuild?view=msvc-170), then configure OCaml with:

```
./configure --host=x86_64-pc-windows CC=clang-cl
```

Having another compilers' opinion on Windows and MSVC (by opposition to MinGW-w64) specific code paths can only benefit the quality of the C code in OCaml. In the future, this could also allow cross-compiling OCaml for native Windows from any host running clang, after installing Windows headers.

This PR accounts for some (in)compatibility quirks of cl and clang-cl CLI, fixes little `configure.ac` bugs introduced in the MSVC restoration effort, adds the required macros to ignore (innocuous) deprecation warnings raised by clang, and adds the only C patch required to make it build out of the box.

### CODE REVIEW

File: testsuite/tests/runtime-C-exceptions/stub_test.c [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#define _CRT_NONSTDC_NO_WARNINGS  /* for strdup */
```
Comment (nojb) [inline testsuite/tests/runtime-C-exceptions/stub_test.c:1]: Nit, but can we put this define _after_ the other `#include`'s?

File: testsuite/tests/embedded/cmstub.c [hunk 2]
```diff
@@ -11,6 +11,7 @@
 /*                                                                     */
 /***********************************************************************/
 
+#define _CRT_NONSTDC_NO_WARNINGS  /* for strdup */
```
Comment (nojb) [inline testsuite/tests/embedded/cmstub.c:14]: Idem.

File: runtime/bigarray.c [hunk 3]
```diff
@@ -132,7 +132,7 @@ static uint16 caml_float_to_float16(float d)
       uint32_t mant_odd = (f.i >> 13) & 1; // resulting mantissa is odd
 
       // update exponent, rounding bias part 1
-      f.i += ((15 - 127) << 23) + 0xfff;
+      f.i += ((uint32_t)(15 - 127) << 23) + 0xfff;
```
Comment (nojb) [inline runtime/bigarray.c]: Do I understand well that this was undefined behaviour? I think this change deserves its own PR to make sure it is not lost in the shuffle. Thanks!

File: configure.ac [hunk 4]
```diff
@@ -851,25 +861,23 @@ AS_CASE([$ocaml_cc_vendor],
...
-    internal_cflags='-experimental:c11atomics -std:c11'
-    CFLAGS="$CFLAGS -experimental:c11atomics -std:c11"
-    OCAML_CL_HAS_VOLATILE_METADATA
-    AS_IF([test "x$cl_has_volatile_metadata" = "xtrue"],
-          [internal_cflags="$internal_cflags -d2VolatileMetadata-"])
-    internal_cppflags="$internal_cppflags -DWINDOWS_UNICODE="
-    internal_cppflags="${internal_cppflags}\$(WINDOWS_UNICODE)"],
+    internal_cflags="$cc_warnings"
+    internal_cppflags="-DUNICODE -D_UNICODE -D_CRT_NONSTDC_NO_WARNINGS \
+-D_WINSOCK_DEPRECATED_NO_WARNINGS"
```
Comment (nojb) [inline configure.ac]: Just to make sure I understand: is `_WINSOCK_DEPRECATED_NO_WARNINGS` needed for the build to succeed, or is it just to silence the warnings. If the eventual objective is to migrate to different functions that do not trigger the warning, we may __not__ want to silence them; what do you think?

**Review by nojb — COMMENTED** [review]
Thanks for this patch!

I did a pass over it, I didn't spot any issues (modulo a few small comments below). There is some tricky `configure` code that may need an expert set of eyes (@shindere, @dra27?) for extra safety. (I'm not approving formally because I didn't review this part of the patch in detail.)

Out of curiosity, do you have a sense of how much work it would be/how feasible it would be to support `clang` (not `clang-cl`) to build on Windows? Thanks!

File: testsuite/tests/runtime-C-exceptions/stub_test.c (hunk 1 above)
Comment (MisterDA) [reply testsuite/tests/runtime-C-exceptions/stub_test.c:1]: No, it needs to be defined before including the system headers. It triggers whether the deprecation warning is attached to the function prototype.
Another option, that is found elsewhere in the code base, would be to write:
```c
#ifdef _WIN32
    return _strdup("bar");
#else
    return strdup("bar");
```

File: configure.ac (hunk 4 above)
Comment (MisterDA) [reply configure.ac]: The define is _solely_ needed to silence the warnings. However as the warnings are turned into errors for us devs, it's then needed for the build to succeed… There's also the option of passing `-Wno-deprecated-declarations` to clang, but the `_WINSOCK_DEPRECATED_NO_WARNINGS` define has the advantage of also working on MSVC. clang warns by default of deprecated functions. We'd need to turn on MSVC warnings with perhaps a high level to trigger them, but that is too noisy currently (I'm somewhat [planning on fixing them](https://github.com/MisterDA/ocaml/pull/41)).
The objective is indeed to move to newer functions (you may peek at https://github.com/MisterDA/ocaml/pull/38). In the meantime, my opinion is that we can move forward using this macro.

File: runtime/bigarray.c (hunk 3 above)
Comment (MisterDA) [reply runtime/bigarray.c]: Ok, but this branch won't build anymore if I remove this commit.

Comment (nojb) [reply runtime/bigarray.c]: Leave it in this PR if you wish, but a separate PR with this change would still be useful, I think, and would be quickly merged anyway.

File: configure.ac (hunk 4 above)
Comment (nojb) [reply configure.ac]: OK, makes sense, thanks.

File: testsuite/tests/runtime-C-exceptions/stub_test.c (hunk 1 above)
Comment (nojb) [reply testsuite/tests/runtime-C-exceptions/stub_test.c:1]: OK, thanks. Since only tests are affected, I think the current solution is OK.

File: runtime/bigarray.c (hunk 3 above)
Comment (MisterDA) [reply runtime/bigarray.c]: Done in #13094.

**Comment by MisterDA** [comment]
> Out of curiosity, do you have a sense of how much work it would be/how feasible it would be to support clang (not clang-cl) to build on Windows? Thanks!

I tried it once, I think the blockers are the same as [ Upgrade from msvcrt.lib to ucrt.lib #122 ](https://github.com/ocaml/flexdll/issues/122), and it would require probably require a new FlexDLL chain.

File: configure.ac [hunk 5]
```diff
@@ -311,7 +311,7 @@ If your host is 64 bits, you can try with './configure CC="gcc -m64"' \
 
 AS_CASE([$host],
   [*-pc-windows],
-    [CC=cl
+    [AS_CASE([$CC], [cl|cl.exe|clang-cl|clang-cl.exe], [], [CC=cl])
```
Comment (shym) [inline configure.ac]: Is there a reason not to set `CC` only if it is empty?
(When I explicitly set some variable, I’m pretty annoyed if that’s overridden)

File: configure.ac [hunk 6]
```diff
@@ -1212,7 +1218,18 @@
 
 # Support for C11 atomic types
 
-OCAML_CC_SUPPORTS_ATOMIC([$cclibs])
+AS_CASE([$ocaml_cc_vendor],
+  [msvc-*],
+    [OCAML_CC_SUPPORTS_ATOMIC
+     AS_IF([! $cc_supports_atomic],
+       [OCAML_CC_SUPPORTS_ATOMIC([-std:c11])
```
Comment (shym) [inline configure.ac]: I would suggest adding `AC_MSG_NOTICE([retrying to detect atomic support with -std:c11])` (and similar for the third attempt) so that the configure log on MSVC doesn't contain three lines in a row checking for the same feature and saying no, no, yes (third time lucky?).

File: configure.ac [hunk 7]
```diff
@@ -942,14 +950,12 @@
...
-       dnl flexdll-sources. This allows us to guarantee that flexdll is always
-       dnl built at one directory level below the root, so we can specify
-       dnl ROOTDIR=..
+       dnl flexdll-sources.
       AS_IF([m4_normalize([test x"$with_flexdll" = 'x'
                             || test x"$with_flexdll" = 'xflexdll'])],
         [AS_IF([test -f 'flexdll/flexdll.h'],
           [flexdll_source_dir=flexdll
-          iflexdir='$(ROOTDIR)/flexdll'
+          iflexdir="$ocamlsrcdir\flexdll"
```
Comment (shym) [inline configure.ac]:
```suggestion
          iflexdir="$ocamlsrcdir\\flexdll"
```

File: configure.ac [hunk 8]
```diff
@@ -851,25 +861,23 @@
     # TODO: see whether the code can be fixed to avoid -Wno-unused
     common_cflags="-O2 -fno-strict-aliasing -fwrapv -mms-bitfields"
     internal_cppflags='-D__USE_MINGW_ANSI_STDIO=0 -DUNICODE -D_UNICODE'
-    internal_cppflags="$internal_cppflags -DWINDOWS_UNICODE="
-    internal_cppflags="${internal_cppflags}\$(WINDOWS_UNICODE)"],
+    internal_cppflags="$internal_cppflags -DWINDOWS_UNICODE=$windows_unicode"],
```
Comment (shym) [inline configure.ac:866]: Do I understand correctly that the problem is in `ocamltest` getting the uninterpreted Makefile variable instead of the value?
An update to `README.win32.adoc:340` to state that it should really be changed in `configure` would be relevant maybe?

**Review by shym — APPROVED** [review]
This looks good to me!
Thank you for cleaning the duplicate flags in MSVC cl calls while you were at it :smile:

I only have a couple of small suggestions:
- I’m a bit confused by the _Autoconf tests won't pick up..._ paragraph in the commit message for _Reorganize MSVC atomics cflags_: I read it as "The flags must be in CFLAGS otherwise the test fails, but we don’t want them to be in CFLAGS". And so, the test should fail???
- I’ve attached the other comments to their corresponding lines.

**Comment by MisterDA** [comment]
I've rebased on top of trunk to get the C fix I extracted from this PR.

Another advantage of clang-cl is that it supports a lot of GCC extensions, such as attributes. clang-cl doesn't masquerade as GNU C (doesn't define `__GNUC__`), but if we switch to detecting attributes support (with `__has_attribute` or C23 `__has_c_attribute`) instead of detecting them based on compiler-specific macros (`_MSC_VER`, `__clang__`, `__GNUC__`, …) we can leverage them on Windows too.

File: configure.ac (hunk 5 above)
Comment (MisterDA) [reply configure.ac]: IIRC the problem is that Autoconf sets `CC` automatically before this line. We have no way of knowing whether the user sets `CC` in `./configure CC=xxx` or if Autoconf discovered the compiler; and Autoconf has a tendency of picking gcc instead of cl or clang-cl. I'll double-check.

File: configure.ac (hunk 6 above)
Comment (MisterDA) [reply configure.ac]: Third time's the charm! I did the three checks because clang-cl doesn't need these flags, we can reasonably expect `-experimental:c11atomics` to disappear in a close future, and `-std:c11` to become unnecessary in a distant one.
I suppose I can instead alter the message in `OCAML_CC_SUPPORTS_ATOMIC` to log which flags it's trying.

File: configure.ac (hunk 7 above)
Comment (MisterDA) [reply configure.ac]: This mirrors similar code for winpthreads:
https://github.com/ocaml/ocaml/blob/e719a1f78c825c1fff4c10c2c47e044149da4caf/configure.ac#L1086

File: configure.ac (hunk 8 above)
Comment (MisterDA) [reply configure.ac:866]: > Do I understand correctly that the problem is in ocamltest getting the uninterpreted Makefile variable instead of the value?

No, it's getting an empty value, `-DWINDOWS_UNICODE=` is what's passed. It's not getting the verbatim text `$(WINDOWS_UNICODE)`. The same of `$(ROOTDIR)` when building FlexDLL.

> An update to [README.win32.adoc:340](https://github.com/ocaml/ocaml/blob/e719a1f78c825c1fff4c10c2c47e044149da4caf/README.win32.adoc?plain=1#L339-L342) to state that it should really be changed in `configure`, maybe?

The original text suggesting of tweaking `Makefile.config` seems fishy to me. The `configure` script is supposed to read `WINDOWS_UNICODE_MODE` from the environment. I'm not sure how everything works here.

**Comment by MisterDA** [comment]
> This looks good to me! Thank you for cleaning the duplicate flags in MSVC cl calls while you were at it 😄

Thanks for the review!

> I’m a bit confused by the _Autoconf tests won't pick up…_ paragraph in the commit message for _Reorganize MSVC atomics cflags_: I read it as "The flags must be in CFLAGS otherwise the test fails, but we don’t want them to be in CFLAGS". 

Yes, that's exactly what's happening. The Autoconf tests uses `CFLAGS` for the compiler invocation, but we don't modify this variable in the `configure` script. If we did, we'd need to be careful not to overwrite user-specified flags (as in `./configure CFLAGS=-myflags`) or end up with duplicated flags here:

https://github.com/ocaml/ocaml/blob/e719a1f78c825c1fff4c10c2c47e044149da4caf/configure.ac#L2597-L2598

but @shindere has a PR removing this behavior: [Don't use configured CFLAGS & CPPFLAGS to compile third-party C sources #12589](https://github.com/ocaml/ocaml/pull/12589).

I do think I've found the sweet spot by passing new `CFLAGS` to `OCAML_CC_SUPPORTS_ATOMIC`. This allows to discover what flags are really needed without modifying the global `CFLAGS` variable; and as no other configure test use C11 atomics or include OCaml headers, they're no risk of a test failing because the compiler invocation would be missing a flag.

I can try to rewrite the commit message if it's not clear enough.

File: configure.ac (hunk 6 above)
Comment (shym) [reply configure.ac]: I had understood your rationale and I agree with it.
An alteration of the message displayed by `OCAML_CC_SUPPORTS_ATOMIC` would be even nicer, indeed.

File: configure.ac (hunk 7 above)
Comment (shym) [reply configure.ac]: :smile:
It flashed in my eyes because there is a version with `\\` just in the next block.
And indeed, the same inconsistency appears in the winpthreads block, shamefully :see_no_evil:

**Comment by shym** [comment]
It was just the phrasing that got me confused then. Would the following sound good?
```
Autoconf tests won't pick up the flags if they're not also in CFLAGS,
otherwise failing atomics support detection. So the necessary flags are
added to CFLAGS but only for the OCAML_CC_SUPPORTS_ATOMIC test,
otherwise they would end up duplicated in the final configuration.
```

File: configure.ac (hunk 5 above)
Comment (MisterDA) [reply configure.ac]: I was mistaken! It's possible to check if `CC` is empty. Good catch, thanks.

File: configure.ac (hunk 6 above)
Comment (MisterDA) [reply configure.ac]: `configure` will now show for `cl`:
```
checking for cl option to support _Atomic types... unsupported
checking for cl option to support _Atomic types... unsupported
checking for cl option to support _Atomic types... -std:c11 -experimental:c11atomics
```
and will show this if the compiler supports C11 out of the box:
```
checking for cl option to support _Atomic types... none needed
```
Still, it's not exactly conveying it's retrying the check with a different set of options, but it's an improvement.

Comment (shym) [reply configure.ac]: Nice! With the small modifications in [this extra commit], I get with cl:

```
checking for cl option to support _Atomic types... failure with no option
checking for cl option to support _Atomic types... failure with -std:c11
checking for cl option to support _Atomic types... -std:c11 -experimental:c11atomics
```

and no change with clang-cl:

```
checking for clang-cl option to support _Atomic types... none needed
```

(I renamed the variable to avoid a line overflow :sweat_smile:)

[this extra commit]: https://github.com/shym/ocaml/commit/543e6decd3ca494b4aac3d6a87b11090db399259

Comment (shym) [reply configure.ac]: And it should be probably better to use a plural:
```diff
-  AC_MSG_CHECKING([for $CC option to support _Atomic types])
+  AC_MSG_CHECKING([for $CC options to support _Atomic types])
```

Comment (MisterDA) [reply configure.ac]: Applied, many thanks!

File: configure.ac (hunk 8 above)
Comment (MisterDA) [reply configure.ac:866]: The documentation is outdated and predates the move to Autoconf. I'll remove the paragraph suggesting to edit `Makefile.config`.

**Comment by nojb** [comment]
@shym: are you happy with the current state of the PR? I am happy to approve it on your behalf based on your quite thorough review.

**Comment by shym** [comment]
@nojb: I just did a last range-diff and I have no further suggestion, I’m happy with the current state of the PR.

**Review by nojb — APPROVED** [review]
LGTM

**Comment by shindere** [comment]
Many thanks to all for the hard work both in authoring and reviewing
htis PR.

I intend to go through it and take care of its merge, which should
preferably happen after the merge of #13095, which I intend to deal
with, too.

File: configure.ac (hunk 7 above)
Comment (dra27) [reply configure.ac]: Sorry to be bringing this in late, but this introduces an unescaped and absolute path into `OC_CPPFLAGS` in `Makefile.build_config.in` which is dangerous (in particular, building in `C:\Devel\ocaml-$(false)` now needs two levels of escaping, but works "correctly" on trunk 😁). If we've ended up with winpthreads also putting an absolute path into the build, really the change ought to be to make it equivalent here (i.e. `$(ROOTDIR)/winpthreads`, if necessary)

Can I check I understand correctly what the clang-cl part of the problem is - we obviously write nonsense into `ocamltest/ocamltest_config.ml` at the moment... is that actually causing an error?

Neither of those two flags should have made it ocamltest's cppflags, as they're specific CPPFLAGS for compiling the runtime, and that feels like a safer fix?

Comment (shym) [reply configure.ac]: Oo, sorry about that.

> we obviously write nonsense into ocamltest/ocamltest_config.ml at the moment... is that actually causing an error?

That was why I went with inlining the path for winpthreads in the MSVC-restoration work, indeed, since quite a few tests were failing when they could not find `pthread.h`. As I hadn’t understood the danger of not using `$(ROOTDIR)`, I didn’t revert to that version after the tests were fixed to use the proper API.

So I just checked that, with the following patch, the MSVC port compiles and tests fine (tested on `trunk`):

```diff
diff --git a/configure.ac b/configure.ac
index fbd8c2e833..53765c9508 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1083,7 +1083,7 @@ AS_IF([test x"$with_winpthreads_msvc" = "xno"],
                           || test x"$with_winpthreads_msvc" = x'winpthreads'])],
       [AS_IF([test -f 'winpthreads/src/winpthread_internal.h'],
         [winpthreads_source_dir=winpthreads
-        iwinpthreadsdir="$ocamlsrcdir\winpthreads\include"],
+        iwinpthreadsdir='winpthreads\include'],
         [AC_MSG_RESULT([required but not available (uninitialized submodule?)])
         AC_MSG_ERROR([exiting])])],
       [rm -rf winpthreads-sources
@@ -1093,7 +1093,7 @@ AS_IF([test x"$with_winpthreads_msvc" = "xno"],
         cp "$with_winpthreads_msvc"/src/*.h winpthreads-sources/src
         cp "$with_winpthreads_msvc"/include/*.h winpthreads-sources/include
         winpthreads_source_dir='winpthreads-sources'
-        iwinpthreadsdir="$ocamlsrcdir\\winpthreads-sources\\include"
+        iwinpthreadsdir='winpthreads-sources\include'
         winpthreadsmsg=" (from $with_winpthreads_msvc)"],
         [AC_MSG_RESULT([requested but not available])
         AC_MSG_ERROR([exiting])])])
@@ -1101,6 +1101,7 @@ AS_IF([test x"$with_winpthreads_msvc" = "xno"],
       [AC_MSG_RESULT([no])],
       [AC_MSG_RESULT([$winpthreads_source_dir$winpthreadmsg])
       OCAML_TEST_WINPTHREADS_PTHREAD_H([$iwinpthreadsdir])
+      iwinpthreadsdir='$(ROOTDIR)\'"$iwinpthreadsdir"
       internal_cppflags="$internal_cppflags -I$iwinpthreadsdir"])],
     [AS_IF([test x"$with_winpthreads_msvc" != 'x'],
       [AC_MSG_RESULT([requested but not supported])
```

> Neither of those two flags should have made it ocamltest's cppflags, as they're specific CPPFLAGS for compiling the runtime, and that feels like a safer fix?

Indeed, that would be much nicer!

**Comment by dra27** [comment]
I had a closer look at things surrounding https://github.com/ocaml/ocaml/pull/13093/commits/55916b0b7d39694ff5fc92c21f8a0d9d7bebe586 and I have a fix for winpthreads in https://github.com/dra27/ocaml/commit/01d9ee555f781a242c35c0d07d74cfce3975e97b and then an overall simplification of all the autoconf-make meta-programming in https://github.com/dra27/ocaml/commit/b88157c1724b9b46a53f61ef3b144c676b8bc631 but then a much simpler fix for the problem with `ocamltest/ocamltest_config.ml` occurred to me... it doesn't need to be using `$internal_cppflags` at all... it's testing the compiler, it both morally and technically ought to be using `$common_cppflags`.

I tried this in https://github.com/dra27/ocaml/commit/12238ffd542f45485f3b1b8bed4a31aa57b89c9e which, when combined with your https://github.com/ocaml/ocaml/pull/13093/commits/2367ceb6e6c43eb85589ced78e70669efd4d77d3 means that `ocamltest_config.ml` now uses both `$common_cflags` and `$common_cppflags`. There are only two tests which use those variables - one of them is simply using the preprocessor for actual pre-processing (it actually doesn't care about any special flags at all) and the other is the `-output-complete-obj` test which doesn't need either of the `-I` or the `-DWINDOWS_UNICODE` flags at all.

So, I think I'd like to suggest dropping https://github.com/ocaml/ocaml/pull/13093/commits/55916b0b7d39694ff5fc92c21f8a0d9d7bebe586 and, if you're happy, then cherry-picking just https://github.com/dra27/ocaml/commit/12238ffd542f45485f3b1b8bed4a31aa57b89c9e. Everything else in this PR looks great - I particularly like the cleanups around `windows_unicode` and the flags detection, thank you!

Once this is in, I'll open a separate PR with the other two commits, because they're orthogonal to this PR, once `ocamltest_config.ml` is being correctly generated.

**Comment by MisterDA** [comment]
> I had a closer look at things […]

Thanks for the review and the help!

> So, I think I'd like to suggest dropping [55916b0](https://github.com/ocaml/ocaml/commit/55916b0b7d39694ff5fc92c21f8a0d9d7bebe586) and, if you're happy, then cherry-picking just [dra27@12238ff](https://github.com/dra27/ocaml/commit/12238ffd542f45485f3b1b8bed4a31aa57b89c9e).

Done.

File: aclocal.m4 [hunk 9]
```diff
@@ -561,9 +548,13 @@ AC_DEFUN([OCAML_QUOTED_STRING_ID], [
...
 AC_DEFUN([OCAML_CC_SUPPORTS_ATOMIC], [
-  AC_MSG_CHECKING([whether the C compiler supports _Atomic types])
-  saved_LIBS="$LIBS"
-  LIBS="$LIBS $1"
+  AC_MSG_CHECKING([for $CC options to support _Atomic types])
+  OCAML_CC_SAVE_VARIABLES
+
+  opts=""
+  AS_IF([test -n "$1"],[CFLAGS="$CFLAGS $1"; opts="$1"])
+  AS_IF([test -n "$2"], [LIBS="$LIBS $2"; opts="${opts:+$opts }$2"])
```
Comment (dra27) [inline aclocal.m4]: Tiny nit:
```suggestion
  AS_IF([test -n "$2"],[LIBS="$LIBS $2"; opts="${opts:+$opts }$2"])
```

File: configure.ac [hunk 10]
```diff
@@ -1212,7 +1222,18 @@ AS_IF([! $arch64],
 
 # Support for C11 atomic types
 
-OCAML_CC_SUPPORTS_ATOMIC([$cclibs])
+AS_CASE([$ocaml_cc_vendor],
+  [msvc-*],
+    [OCAML_CC_SUPPORTS_ATOMIC
+     AS_IF([! $cc_supports_atomic],
```
Comment (dra27) [inline configure.ac]: A thought: we could keep the line above (well, the `OCAML_CC_SUPPORS_ATOMIC([], [$cclibs])` version from below), which is the line that ultimately we expect to work on all platforms. Then:
```suggestion
AS_CASE([$cc_supports_atomic,$ocaml_cc_vendor],
  [false,msvc-*],
    [AS_IF([! $cc_supports_atomic],
```
(plus bracket adjustments)

File: aclocal.m4 [hunk 11]
```diff
@@ -561,9 +548,13 @@ AC_DEFUN([OCAML_QUOTED_STRING_ID], [
 ])
 
 AC_DEFUN([OCAML_CC_SUPPORTS_ATOMIC], [
-  AC_MSG_CHECKING([whether the C compiler supports _Atomic types])
-  saved_LIBS="$LIBS"
-  LIBS="$LIBS $1"
+  AC_MSG_CHECKING([for $CC options to support _Atomic types])
```
Comment (dra27) [inline aclocal.m4]: Sorry to bike-shed this slightly further, but I think it's slightly more autoconf-y to include the option being tried here, and have the result still be yes/no (see similar checking for pthreads in build-aux/ax_pthread.m4); so (with this line moved below the computation of `$opts`, obviously):
```suggestion
  AC_MSG_CHECKING([if $CC support _Atomic types with ${opts:-no additional options}
```

File: configure.ac (hunk 4 above)
Comment (dra27) [reply configure.ac]: Are there sufficiently few places that it could be done with `#define _WINSOCK_DEPRECATED_NO_WARNINGS /* Remove when gethostbyname replaced with getaddrinfo */` in the files? (I'm not 100% convinced this would be better)

**Review by dra27 — COMMENTED** [review]
Fully reviewed, and looks good to go, thanks - a mildly bike-sheddy suggestion on the atomics flags detection

File: aclocal.m4 (hunk 11 above)
Comment (MisterDA) [reply aclocal.m4]: Ok! I tried to make mine look like `checking for CC options to support C11` and related, but I agree that printing options is nicer.

**Comment by MisterDA** [comment]
> Are there sufficiently few places that it could be done with `#define _WINSOCK_DEPRECATED_NO_WARNINGS /* Remove when gethostbyname replaced with getaddrinfo */` in the files? (I'm not 100% convinced this would be better)

Yes, the problem happens only in `runtime/debugger.c`. I'll move the define there. I already have written the patches to update this code, and reverting this simple commit will be even simpler.

EDIT: my bad, it's also needed in libunix, of course. We'll need to keep the macro there because we still want to provide bindings for deprecated functions.

**Comment by MisterDA** [comment]
Many thanks to all the reviewers! I think I've taken everything into account, and I've rebased the PR on trunk. All set?

**Comment by shindere** [comment]
Antonin Décimo (2024/04/22 10:40 -0700):
> Many thanks to all the reviewers! I think I've taken everything into
> account, and I've rebased the PR on trunk. All set?

I think we are (almost?) there!

I have a few neits about the unicode commit but they may be worth left
for another day, after all.

Many thanks to you, @MisterDA, for the super work you have done here.

I also thank the other reviewers for their very nice reviews!

**Comment by shindere** [comment]
Merging, as thit opens the way for a few already-planned subsequent improvements.
